include/deal.II/base/thread_local_storage.h

```diff
@@ -265,6 +308,36 @@ namespace Threads
   {}
 
 
+  template <typename T>
+  inline ThreadLocalStorage<T> &
+  ThreadLocalStorage<T>::operator=(const ThreadLocalStorage<T> &t)
+  {
+    /*
+     * We need to raise the reader lock of the argument and our writer lock
+     * while copying internal data structures.
+     */
+    std::shared_lock<decltype(insertion_mutex)> reader_lock(t.insertion_mutex);
+    std::unique_lock<decltype(insertion_mutex)> writer_lock(insertion_mutex);
+
+    data = t.data;
+  }
+
+
+  template <typename T>
+  inline ThreadLocalStorage<T> &
+  ThreadLocalStorage<T>::operator=(ThreadLocalStorage<T> &&t)
+  {
+    /*
+     * We need to raise the reader lock of the argument and our writer lock
+     * while copying internal data structures.
+     */
+    std::shared_lock<decltype(insertion_mutex)> reader_lock(t.insertion_mutex);
```

Comment:
@bangerth If you happen to keep the move constructors/assignment operators: I made a small semantic mistake, the `reader_lock` should also be a `std::unique_lock`, aka a writer lock, as well. But that's more or less a moot point (the `t.data` object is invalid after the move and so is any access to the ThreadLocalStorage object `t`).